pdata/internal/generated_immutable_slice.go

```diff
@@ -0,0 +1,132 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Code generated by "model/internal/cmd/pdatagen/main.go". DO NOT EDIT.
+// To regenerate this file run "go run model/internal/cmd/pdatagen/main.go".
+
+package internal
+
+// ImmutableByteSlice represents a []byte slice that cannot be mutated.
+type ImmutableByteSlice struct {
+	value []byte
+}
+
+// NewImmutableByteSlice creates a new ImmutableByteSlice by moving an underlying array of the provided []byte slice.
+// Note that the argument value will be cleared after the call,
+// use NewImmutableByteSliceFromCopy if it needs to be preserved.
+func NewImmutableByteSlice(val *[]byte) ImmutableByteSlice {
+	is := ImmutableByteSlice{}
+	if val != nil {
+		is.value = *val
+		*val = nil
+	}
+	return is
+}
+
+// NewImmutableByteSliceFromCopy creates a new ImmutableByteSlice by copying the provided []byte slice.
+// Use NewImmutableByteSlice if the provided slice is not needed anymore.
+func NewImmutableByteSliceFromCopy(val []byte) ImmutableByteSlice {
```

Comment:
Makes sense, signatures of `CopyFrom` and `MoveFrom` will be different from `CopyTo` and `MoveTo` that we have but I think that's ok. `CopyFromRaw` or `CopyFromBytes` probably to verbose

Reply:
Maybe `CopyRaw` and `MoveRaw` as another option?

Reply:
We also have getter method called `Value`. So setters can use the term. For example:
- `Value() []byte`, `MoveValue(val *[]byte)`, `CopyValue(val []byte)`
- `Raw() []byte`, `MoveRaw(val *[]byte)`, `CopyRaw(val []byte)`

Reply:
I tried this approach at it seems like the interface becomes clunky. In order to use a new `ImmutableByteSlice` user would have to write 3 lines:
```golang
ibs := ImmutableByteSlice()
ibs.MoveFrom(&[]{1,2,3})
use(ibs)
```

instead of

```golang
use(NewImmutableByteSlice(&[]{1,2,3}))
```

Unless we `MoveValue` and `CopyValue` returns the `ImmutableByteSlice` instance itself which doesn't seem ideal.

Also use case when the same instance can be used for several calls of `MoveFrom` or `CopyFrom` seems unrealistic.

What about `NewImmutableByteSliceFromValue` and `NewImmutableByteSliceFromCopy`?

Reply:
> Also use case when the same instance can be used for several calls of MoveFrom or CopyFrom seems unrealistic.

Not really, for histogram bucket boundaries will lots of times be the case correct?

Reply:
I'm going to apply these changes in contrib and will see how it looks

Reply:
@bogdandrutu I applied the changed in config and didn't run into use cases where the `Copy/Move` functionality would be preferable. https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9940